### ISSUE TITLE
Use option as quick entity value

### ISSUE THREAD

**Issue by bennett-sh** [atampy25/simple-mod-framework#535]
<strong>Please tell me your opinion about this idea because I will probably implement it myself to some degree and make a PR.</strong>

### What?
I think it would be great if you could add placeholders for options in QuickEntity/repo files so you wouldn't have to create multiple files for one option. You could possibly even make it possible to create a template-like file (example 2).

### Why?
- Quicker mod developement (you can use one single content folder)
- New possibilities:
  - Pass strings, numbers and more to mods (example usecase below)
  - Deeper user customisation of mods and more "high-level" mods
    - High-level mod example: change name of any NPC (example 2)

### Implementation
The functionality in example 1 could easily be recreated by processing the JSON file before deploy. Example 2 would be more difficult to implement but you could make a new file type (```.something.entity.json``` and ```.something.repository.json```) and replace the arbitrary values `(SMF_OPTIONS:...)` with JSON values like `"(SMF_OPTIONS:...)"` to make the JSON parsable.

### Example
This example would let you rename an NPC of your choice to any name of your choice and add an option to allow firing inside the safehouse.

`content/chunk29/safehouse.patch.json`:
```json
{
  "tempHash": "00A9B93A825F5BDC",
  "tbluHash": "005BA5CA6AC00627",
  "patch": [
    {
      "SubEntityOperation": ["e07c81c25ad49639", { "SetPropertyValue": { "property_name": "m_bFire", "value": (SMF_OPTION:allowFire) } }]
    }
  ],
  "patchVersion": 6
}
```

`content/chunk0/npc.repository.json`:
```json
{
  (SMF_OPTION:npcID): {
    "Name": (SMF_OPTION:npcName)
  }
}
```

`manifest.json`:
```json
{
  "options": [
    {
      "name": "Allow Fire",
      "tooltip": "Tick to allow shooting.",
      "placeholderName": "allowFire",
      "type": "checkbox"
    },
    {
      "name": "NPC Name",
      "tooltip": "Use this option to change the name of an NPC.",
      "placeholderName": "npcName",
      "type": "text"
    },
    {
      "name": "Alicia Rogers",
      "group": "NPC ID",
      "placeholderName": "npcID",
      "placeholderValue": "5c5ec41e-b4c6-43d4-a2df-364a4f340dcc",
      "type": "select"
    },
    {
      "name": "Maggie Fan",
      "group": "NPC ID",
      "placeholderName": "npcID",
      "placeholderValue": "13af00e6-80e9-46d4-bffd-a2b4c71e6262",
      "type": "select"
    }
  ]
}
```

#### (Example 2)
`content/chunk0/npc.something.repository.json`:
```json
{
  (SMF_OPTION:npcID): {
    "Name": (SMF_OPTION:npcName)
  }
}
```

`manifest.json`:
```json
{
  "options": [
    {
      "name": "Change NPC Names",
      "type": "map",
      "key": { "type": "text", "name": "npcID" },
      "value": { "type": "text", "name": "npcName" }
    }
  ]
}
```

This would turn the options
13af00e6-80e9-46d4-bffd-a2b4c71e6262 -> "My NPC 1"
a432cb53-d7ef-4724-a2ba-27a25b6f39ee -> "My NPC 2"
into the following repository JSON:
```json
{
  "13af00e6-80e9-46d4-bffd-a2b4c71e6262": {
    "Name": "My NPC 1"
  },
  "a432cb53-d7ef-4724-a2ba-27a25b6f39ee": {
    "Name": "My NPC 2"
  }
}
```

**Comment by atampy25** [comment]
I did consider this as part of the original option design (after all, it does have value); it was rejected for two reasons. The first is purely performance; having to scan every file for arbitrary tags to replace would likely decrease performance (maybe not significantly but still), as well as requiring a cache invalidation on any changes. The second is these tags are "magic" in a way that the framework generally tries to avoid; there isn't a clear connection between these tags and their options in the same way as there is with other features (adding onto that, the "parseable" tags would have to be used because tools like QNE would require a complete revamp to work with the arbitrary strings; this also causes concerns because a `"(option:bla)"` would ordinarily be considered a string, and would be assumed to be a string by most users, but would have to be reprocessed to a boolean if that was the type of the option).

Your first example (allowing shooting in the safehouse) also brings up an issue. That patch could be re-worked into a single content folder for allowing shooting and then a `checkbox` option that enables or disables it (since the default is no shooting). This might create a slightly more tedious experience since you have to create two folders instead of adding a tag, but think about what happens if that option is disabled - the JSON is reprocessed and the QN patch sets it to false (which is the default). The framework would then apply the patch, requiring RPKG import (deploy time penalty) and a hash probe for both the TEMP and TBLU. It would also have to spin up a patch worker, which requires another RPKG instance, to apply the (useless) patch. Realistically, with the performance improvements the framework has gained over time, this probably wouldn't be too much of  a performance issue. But I don't want to encourage adding effectively useless patches to mods purely because of a slight increase in developer ergonomics - mod developers already tend to prioritise their experience over that of users even in cases of huge performance increases from running a single BAT file.

This functionality is certainly possible to manually implement using a mod script - a simple `fs.readFileSync("bla.json", "utf8").replaceAll("(option:bla)", <check if option is checked>)` could be staged as virtual content. The only barrier there is the connection to the UI, but that's on the [Trello](https://trello.com/c/3uq1KHCs/29-scripting-based-options).

TL;DR - I don't think it's a good idea because it doesn't follow the established philosophy for framework features; modders can however implement it using mod scripts, and I may work on an example script displaying this potential.

**Comment by bennett-sh** [comment]
Thanks for the overview. I didn't know mod scripts could do that so I guess
the only thing still "required" to get the full functionality is the option
to use numbers & text inputs.
